x/ibc-transfer/types/trace.go

```diff
@@ -0,0 +1,196 @@
+package types
+
+import (
+	"encoding/hex"
+	"errors"
+	fmt "fmt"
+	"strings"
+
+	"github.com/tendermint/tendermint/crypto/tmhash"
+	tmbytes "github.com/tendermint/tendermint/libs/bytes"
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	host "github.com/cosmos/cosmos-sdk/x/ibc/24-host"
+)
+
+// ParseDenomTrace parses a string with the ibc prefix (denom trace) and the base denomination
+// into a DenomTrace type.
+//
+// Examples:
+//
+// 	- "portidone/channelidone/uatom" => DenomTrace{Trace: "portidone/channelidone", BaseDenom: "uatom"}
+// 	- "uatom" => DenomTrace{Trace: "", BaseDenom: "uatom"}
+func ParseDenomTrace(rawDenom string) DenomTrace {
+	denomSplit := strings.Split(rawDenom, "/")
+
+	if denomSplit[0] == rawDenom {
+		return DenomTrace{
+			Trace:     "",
+			BaseDenom: rawDenom,
+		}
+	}
+
+	return DenomTrace{
+		Trace:     strings.Join(denomSplit[:len(denomSplit)-1], "/"),
+		BaseDenom: denomSplit[len(denomSplit)-1],
+	}
+}
+
+// Hash returns the hex bytes of the SHA256 hash of the DenomTrace fields using the following formula:
+//
+// hash = sha256(trace + "/" + baseDenom)
+func (dt DenomTrace) Hash() tmbytes.HexBytes {
+	return tmhash.Sum([]byte(dt.GetPrefix() + dt.BaseDenom))
+}
+
+// GetPrefix returns the receiving denomination prefix composed by the trace info and a separator.
+func (dt DenomTrace) GetPrefix() string {
+	return dt.Trace + "/"
+}
+
+// IBCDenom a coin denomination for an ICS20 fungible token in the format 'ibc/{hash(trace +
+// baseDenom)}'. If the trace is empty, it will return the base denomination.
+func (dt DenomTrace) IBCDenom() string {
+	if dt.Trace != "" {
+		return fmt.Sprintf("ibc/%s", dt.Hash())
+	}
+	return dt.BaseDenom
+}
+
+// RemovePrefix trims the first portID/channelID pair from the trace info. If the trace is already
+// empty it will perform a no-op. If the trace is incorrectly constructed or doesn't have separators
+// it will return an error.
+func (dt *DenomTrace) RemovePrefix() {
+	if dt.Trace == "" {
+		return
+	}
+
+	traceSplit := strings.SplitN(dt.Trace, "/", 3)
+
+	switch {
+	// NOTE: other cases are checked during msg validation
+	case len(traceSplit) == 2:
+		dt.Trace = ""
+	case len(traceSplit) == 3:
+		dt.Trace = traceSplit[2]
+	}
+}
+
+func validateTraceIdentifiers(identifiers []string) error {
+	if len(identifiers) == 0 || len(identifiers)%2 != 0 {
+		return errors.New("trace info must come in pairs of port and channel identifiers '{portID}/{channelID}'")
+	}
+
+	// validate correctness of port and channel identifiers
+	for i := 0; i < len(identifiers); i += 2 {
```

Comment:
from what I remember, port and channel identifier verification is an implementation decision (but maybe this is changing). If this is true, I don't think it makes much sense to impose the SDK implementation restrictions on port and channel identifiers we aren't connection with.

Thoughts @cwgoes ?

Reply:
At the moment, ICS 24 specifies length minimums & maximums along with character restrictions:

https://github.com/cosmos/ics/tree/master/spec/ics-024-host-requirements#paths-identifiers-separators

Are we introducing additional constraints here? If so, we should either remove them or update the spec.

Reply:
we are using the default ones from the spec